src/commands/build-util.ts

```diff
@@ -1,19 +1,112 @@
 import type CSSModuleLoader from 'css-modules-loader-core';
-import type {EventEmitter} from 'events';
-import execa from 'execa';
-import npmRunPath from 'npm-run-path';
-import {
-  BuildScript,
-  SnowpackConfig,
-  SnowpackPluginBuildArgs,
-  SnowpackPluginBuildResult,
-} from '../config';
+import {EventEmitter} from 'events';
+import path from 'path';
+import {SnowpackBuildMap, SnowpackConfig, SnowpackPlugin, SnowpackSourceFile} from '../config';
+import {getExt, replaceExt} from '../util';
+
+export interface BuildFileOptions {
+  buildPipeline: SnowpackPlugin[];
+  messageBus: EventEmitter;
+  isDev: boolean;
+}
 
 const IS_PREACT = /from\s+['"]preact['"]/;
 export function checkIsPreact(filePath: string, contents: string) {
   return filePath.endsWith('.jsx') && IS_PREACT.test(contents);
 }
 
+export function getInputsFromOutput(fileLoc: string, plugins: SnowpackPlugin[]) {
+  const ext = path.extname(fileLoc);
+  const potentialInputs = new Set([fileLoc]);
+  for (const plugin of plugins) {
+    if (plugin.output.includes(ext)) {
+      plugin.input.forEach((inp) => potentialInputs.add(fileLoc.replace(ext, inp)));
+    }
+  }
+  return Array.from(potentialInputs);
+}
+
+/** Core Snowpack file pipeline builder */
+export async function buildFile(
+  srcPath: string,
+  srcContents: string,
+  {buildPipeline, messageBus, isDev}: BuildFileOptions,
+): Promise<SnowpackBuildMap> {
+  const srcExt = getExt(srcPath);
+  const output: SnowpackBuildMap = {}; // important: clear output for each src file to keep memory low
+  output[srcPath] = {...srcExt, contents: srcContents, locOnDisk: srcPath}; // this is the object we’ll mutate through transformations
+
+  // transform with Snowpack plugins (main build pipeline)
+  for (const step of buildPipeline) {
+    const staleOutputs = new Set<string>();
+    for (const destPath of Object.keys(output)) {
+      const startBuildFile = output[destPath];
+      if (!step.input.includes(getExt(destPath).baseExt)) {
```

Comment:
What is this doing? I can see that some things skip building, but not 100% sure on the conditions/intent.

Reply:
+1 adding a comment